prow/plugins/assign/assign.go

```diff
@@ -104,58 +127,155 @@ func parseLogins(text string) []string {
+func createCommentResponse(gc githubClient, e event, msg string) error {
+	if err := gc.CreateComment(e.org, e.repo, e.number, plugins.FormatResponseRaw(e.body, e.url, e.login, msg)); err != nil {
+		return fmt.Errorf("comment err: %v", err)
 	}
+	return nil
+}
 
-	matches := assignRe.FindAllStringSubmatch(ae.body, -1)
+func combineErrors(err1, err2 error) error {
+	if err1 != nil && err2 != nil {
+		return fmt.Errorf("Two errors: 1)%v 2)%v.", err1, err2)
+	} else if err1 != nil {
+		return err1
+	} else {
+		return err2
+	}
+}
+
+// handle is the generic handler for the assign plugin. It uses the handler's regexp and affectedLogins
+// functions to identify the users to add and/or remove and then passes the appropriate users to the
+// handler's add and remove functions. If add failes to add some of the users, the handler's
+// handleAddFailure function is called.
+func handle(h handler) error {
+	e := h.event()
+	if e.action != "created" && e.action != "opened" { // Only consider new comments/issues
+		return nil
+	}
+	matches := h.regexp().FindAllStringSubmatch(e.body, -1)
 	if matches == nil {
 		return nil
 	}
-
-	assignments := make(map[string]bool)
+	users := make(map[string]bool)
 	for _, re := range matches {
-
-		add := re[1] != "un" // unassign == !add
-		if re[2] == "" {
-			assignments[ae.login] = add
-			continue
-		}
-		for _, login := range parseLogins(re[2]) {
-			assignments[login] = add
+		add := re[1] != "un" // un<cmd> == !add
+		for _, login := range h.affectedLogins(re[2]) {
+			users[login] = add
 		}
 	}
-
-	var assign, unassign []string
-
-	for login, add := range assignments {
+	var toAdd, toRemove []string
+	for login, add := range users {
 		if add {
-			assign = append(assign, login)
+			toAdd = append(toAdd, login)
 		} else {
-			unassign = append(unassign, login)
+			toRemove = append(toRemove, login)
 		}
 	}
 
-	if len(unassign) > 0 {
-		log.Printf("Removing assignees from %s/%s#%d: %v", ae.org, ae.repo, ae.number, unassign)
-		if err := gc.UnassignIssue(ae.org, ae.repo, ae.number, unassign); err != nil {
+	if len(toRemove) > 0 {
+		if err := h.remove(toRemove); err != nil {
 			return err
 		}
 	}
-	if len(assign) > 0 {
-		log.Printf("Adding assignees to %s/%s#%d: %v", ae.org, ae.repo, ae.number, assign)
-		if err := gc.AssignIssue(ae.org, ae.repo, ae.number, assign); err != nil {
+	if len(toAdd) > 0 {
+		if err := h.add(toAdd); err != nil {
 			if mu, ok := err.(github.MissingUsers); ok {
-				msg := fmt.Sprintf("GitHub didn't allow me to assign the following users: %s.\n\nNote that only [%s members](https://github.com/orgs/%s/people) can be assigned.",
-					strings.Join(mu, ", "), ae.org, ae.org)
-				if e2 := gc.CreateComment(ae.org, ae.repo, ae.number, plugins.FormatResponseRaw(ae.body, ae.url, ae.login, msg)); e2 != nil {
-					return fmt.Errorf("comment err: %v", e2)
-				}
-				return nil
+				return h.handleAddFailure(mu)
 			}
 			return err
 		}
 	}
 	return nil
 }
+
+type handler interface {
```

Comment:
I think you can do away with the interface and go with a single struct that contains those functions. Note that the GitHub client function signatures for the two implementations of the functions are identical, so RequestReview looks just like AssignIssue if you squint at it.